scipy/optimize/_sda.py

```diff
@@ -0,0 +1,572 @@
+                self.state.xbest = np.copy(x)
+                self.state.current_energy = e
+                self.state.current_location = np.copy(x)
+                return
+        # Check probability of a need to perform a LS even if no improvment
+        # (Dual annealing principle)
+        do_ls = False
+        if self.K < 90 * len(self.state.current_location):
+            pls = np.exp(self.K * (self.state.ebest - self.state.current_energy
+                                   ) / self.temperature_step)
+            if pls >= self._rs.random_sample():
+                do_ls = True
+        # Global energy not improved, let's see what LS gives
+        # on the best Markov chain location
+        if self.not_improved_idx >= self.not_improved_max_idx:
+            do_ls = True
+        if do_ls:
+            e, x = self.ofw.local_search(self.xmin)
+            self.xmin = np.copy(x)
+            self.emin = e
+            self.not_improved_idx = 0
+            self.not_improved_max_idx = self.state.current_location.size
+            if e < self.state.ebest:
+                self.state.ebest = self.emin
+                self.state.xbest = np.copy(self.xmin)
+                self.state.current_energy = e
+                self.state.current_location = np.copy(x)
+
+
+class ObjectiveFunWrapper(object):
+    """
+    Class used to wrap around the objective function in order to apply local
+    search and default gradient computation.
+    Default local minimizer is L-BFGS-B
+    """
+    def __init__(self, bounds, func, **kwargs):
+        self.func = func
+        self.nb_fun_call = 0
+        self.kwargs = kwargs
+        self.minimizer = minimize
+        self.fun_args = None
+        lu = list(zip(*bounds))
+        self.lower = np.array(lu[0])
+        self.upper = np.array(lu[1])
+        self.ls_max_iter = self.lower.size * 6
+
+        if self.ls_max_iter < 100:
+            self.ls_max_iter = 100
+        if self.ls_max_iter > 1000:
+            self.ls_max_iter = 1000
+
+        # By default, scipy L-BFGS-B is used with a custom 3 points gradient
+        # computation
+        else:
+            self.fun_args = ()
+        if not self.kwargs or 'method' not in self.kwargs:
+            self.kwargs['method'] = 'L-BFGS-B'
+            self.kwargs['options'] = {
+                'disp': None, 'maxls': 100, 'iprint': -1, 'gtol': 1e-06,
+                'eps': 1e-06,
+                'maxiter': self.ls_max_iter,
+                'maxcor': 10, 'maxfun': 15000
+            }
+            if 'jac' not in self.kwargs:
+                self.kwargs['jac'] = self.gradient
+            if 'bounds' not in self.kwargs:
+                self.kwargs['bounds'] = bounds
+        if 'args' in self.kwargs:
+            self.fun_args = self.kwargs.get('args')
+
+    def func_wrapper(self, x):
+        self.nb_fun_call += 1
+        return self.func(x, *self.fun_args)
+
+    def gradient(self, x):
+        g = approx_derivative(self.func_wrapper, x)
```

Comment:
I just take a look at the internals of L-BFGS-B function, and it seems like the FORTRAN implementation already has an internal implementation of finite-differences. I think it would be better to use it rather than the SciPy implementation (as I have suggested at first). I think it is more elegant, since you delegate the task of approximating derivatives entirely to ``minimize``.

To use it you only need to omit the gradient as argument to ``minimize``.

Reply:
I have made some tests and it seems that the default finite-differences calculation is less accurate than the SciPy approx_derivative (which uses 3-points by default). Therefore, I would suggest that for the default sda local minimizer, it uses the approx_derivative as you suggested first.

Reply:
Thank you very much for digging into this @sgubianpm. I think there are two points we should consider:

1. I think we should try to make the function ``sda`` as independent as possible from ``minimize``. And delegating the derivatives computation entirely to minimize seems like a good practice.
2. 2-points vs 3-point derivatives approximation is not such a simple decision. While 3-point is more accurate it also demands more functions evaluations. So it is not necessarily better.

Anyway, I don't think the internal implementation is necessarily better, but I think using it would result in a more modular and clean implementation.

Reply:
Ok, I'll remove the custom gradient and let minimize do the job.

Reply:
Great :)